# All about **Strings** in Python

---
## Introduction and Basics
A string is a sequence of characters. Anything entered inside quotes is considered a string in Python.

In [1]:
print('Hi there! How are you doing?')

Hi there! How are you doing?


In [2]:
text = " This is a sample text stored in a variable"
print(text)

 This is a sample text stored in a variable


In [3]:
#to print quotes as output
print("Today's a new Day!")

Today's a new Day!


In [4]:
print('Today\'s a new day')

Today's a new day


In [5]:
print('click "OK"')

click "OK"


In [6]:
print("Press \"Enter\"")

Press "Enter"


In [7]:
t1 = 'text 1'
t2 = "text 2"
print (t1, t2)

text 1 text 2


In [8]:
print(t1 + t2)

text 1text 2


In [9]:
print(t1 * 3)

text 1text 1text 1


#### Multi-line Strings

In [10]:
print(''' You can
type a multi-line
string like
this using any quotation marks
three times.''')

 You can
type a multi-line
string like
this using any quotation marks
three times.


#### Escape Characters

##### using \n and \t
The \n is used to go to new line and the \t is used for creating tab space. These are called **Escape Characters**.

In [11]:
print("another way to type \n a multi-line statement \n is by this way")

another way to type 
 a multi-line statement 
 is by this way


In [12]:
print('testing \ttab space \toption here')

testing 	tab space 	option here


#### Printing Escape Characters
To print escape characters, there are a number of ways

In [13]:
print("This \nis a \tsample text")

This 
is a 	sample text


**Using Raw Strings**

In [14]:
print(r"This \nis a \tsample text")

This \nis a \tsample text


**Using repr()**

In [15]:
print(repr("This \nis a \tsample text"))

'This \nis a \tsample text'


**Using double backslashes**

In [16]:
print("This \\nis a \\tsample text")

This \nis a \tsample text


There many other ways to do this but for now this is enough....

#### end Parameter in Python `print`
The ```end``` parameter in Python's built-in print() function controls what character or string is appended at the end of the printed output.\
By default, the end parameter is set to '\n' (the newline character), which means that after each print() statement executes, a new line is automatically inserted, causing subsequent output to appear on the next line.\
You can customize this behavior by providing a different string value to the end parameter. This allows you to:
- **Print on the same line:**\
Set end=' ' to add a space instead of a newline, keeping subsequent output on the same line.
- **Add custom characters:**\
Use end='@' or any other string to append specific characters after the printed content.
- **Print nothing at the end:**\
Set end='' to prevent any character from being added at the end of the line.

In [17]:
# Ending with a space
print("Hello", end=' ')
print("World")

Hello World


In [18]:
# Ending with a custom string
print("Python", end='***')
print("Programming")

Python***Programming


In [19]:
for i in range(5):
    print(i, end=", ")   # this is a 'For loop' will get to it in further sections...

0, 1, 2, 3, 4, 

#### sep Parameter in Python `print`
The ```sep``` parameter within the print() function specifies the separator between the values being printed.\
By default, sep is set to a single space character (' '). \
This means that when you print multiple items, they will be separated by a space. 

In [20]:
# Default behavior (sep=' ' is implied)
print("Hello", "World")

Hello World


In [21]:
# Using a custom separator: a dash
print("apple", "banana", "cherry", sep="-")

apple-banana-cherry


In [22]:
# Using an empty string as a separator (no space)
print("P", "A", "I", sep="")

PAI


In [23]:
# Using a more complex separator
name = "alex"
account = "gmail.com"
print(name, account, sep='@')

alex@gmail.com


---
## String Comparison in Python
Python supports several operators for string operators as follows:\
== check for Equality \
!= check for Inequality \
<, > , <=, >= perform lexicographical comparisons based on Alphabetical order

In [24]:
s1 = "Python"
s2 = "Python"
s3 = "pyhton"
s4 = "Java"

In [25]:
print(s1 == s2)

True


In [26]:
print(s1 == s3)

False


In [27]:
print(s1 != s4)

True


In [28]:
print(s2 != s3)

True


#### Case-Insensitive Comparison

In [29]:
print(s2.lower() == s3.lower())

False


#### Lexicographical Comparison

In [30]:
f1 = "apple"
f2 = "banana"
f3 = "fig"

In [31]:
print(f1 < f2)

True


In [32]:
print(f3 > f1)

True


---
## Python String Methods
Some of the most commonly used String methods are as follows:

#### Case related String Methods
**`lower` `upper` `title` `capitalize` `swapcase`**

In [33]:
line = 'This is a Sample RANdom TexT.'

In [34]:
print(line)

This is a Sample RANdom TexT.


In [35]:
print(line.lower())

this is a sample random text.


In [36]:
print(line.upper())

THIS IS A SAMPLE RANDOM TEXT.


In [37]:
print(line.title())

This Is A Sample Random Text.


In [38]:
print(line.capitalize())

This is a sample random text.


In [39]:
print(line.swapcase())

tHIS IS A sAMPLE ranDOM tEXt.


#### More commonly used String Methods

In [40]:
sent = '     This is another sample sentence to check functions on.   '

In [41]:
print(sent)

     This is another sample sentence to check functions on.   


**`strip`**

In [42]:
newsent = sent.strip()

In [43]:
print(newsent)

This is another sample sentence to check functions on.


**`find`**

In [44]:
print(sent.find('sample'))
print(newsent.find('sample'))

21
16


**`count`**

In [45]:
print(newsent.count('s'))
print(newsent.count('n'))

5
6


**`len`**

In [46]:
print(len(sent))
print(len(newsent))

62
54


**`replace`**

In [47]:
print(newsent.replace('sample','new'))

This is another new sentence to check functions on.


In [48]:
print(newsent.replace('sample','new').count('n'))

7


**`split`**

In [49]:
sent_list = newsent.split(' ')
print(sent_list)

['This', 'is', 'another', 'sample', 'sentence', 'to', 'check', 'functions', 'on.']


**`join`**

In [50]:
j1_sent = ' '.join(sent_list)
j2_sent = '-'.join(sent_list)
j3_sent = '|'.join(sent_list)
j4_sent = '_|_'.join(sent_list)

In [51]:
print(j1_sent, '\n')
print(j2_sent, '\n')
print(j3_sent, '\n')
print(j4_sent, '\n')

This is another sample sentence to check functions on. 

This-is-another-sample-sentence-to-check-functions-on. 

This|is|another|sample|sentence|to|check|functions|on. 

This_|_is_|_another_|_sample_|_sentence_|_to_|_check_|_functions_|_on. 



---
## Indexing & Slicing Strings in Python

In [52]:
name = "Aliya Fanaskar"

In [53]:
print(len(name))

14


**`  A    l    i    y    a         F    a    n    a    s    k    a    r`**\
`  0    1    2    3    4    5    6    7    8    9    10   11   12   13`\
`-14  -13  -12   -11  -10   -9   -8   -7   -6   -5   -4   -3   -2   -1`


list slicing syntax: `list[start:end:step]`

In [54]:
name[0]

'A'

In [55]:
name[1]

'l'

In [56]:
name[-1]

'r'

In [57]:
name[-7]

'a'

In [58]:
name[1:5]

'liya'

In [59]:
name[-4:-1]

'ska'

In [60]:
name[9:3]

''

In [61]:
name[-6:-12]

''

In [62]:
name[:7]

'Aliya F'

In [63]:
name[:-10]

'Aliy'

In [64]:
name[3:]

'ya Fanaskar'

In [65]:
name[-6:]

'naskar'

In [66]:
name[::-1]

'raksanaF ayilA'

In [67]:
name[::2]

'AiaFnsa'

In [68]:
name[1::2]

'ly aakr'

In [69]:
name[2:12:3]

'i nk'

## String Membership Testing 

#### Using in Method

In [70]:
f = 'Words for Fun'

In [71]:
print("words" in f)

False


In [72]:
print("Words" in f)

True


#### Using startswith() and endswith() Method

In [73]:
s = "Hello world"

In [74]:
print(s.startswith('Hello'))

True


In [75]:
print(s.endswith('World'))

False


In [76]:
print(s.endswith('world'))

True


## String Immutability 
Strings are immutable meaning that they cannot be changed after they are created.

In [77]:
t = "Sample Text"

In [78]:
#Trying to change the first character raises an error

In [79]:
t[0]

'S'

In [80]:
t[1:]

'ample Text'

In [81]:
# t[0] = 'T'         # will give TypeError

In [82]:
s = "Tr" + t[1:]
print(s)

Trample Text


---
## Formatting Python String

#### Using f-strings

In [83]:
name = "Aliya"
age = 25
print(f"Name: {name}, Age: {age}")
print(f"Message from Candidate: Hi, my name is {name} and I'm {age} years old.")

Name: Aliya, Age: 25
Message from Candidate: Hi, my name is Aliya and I'm 25 years old.


In [84]:
a = 15
b = 20
print(f"the sum of {a} and {b} is {a+b}")

the sum of 15 and 20 is 35


In [85]:
print(f"{{Hello there!}}")

{Hello there!}


In [86]:
print(f"{{{{{{Hello there!}}}}}}")

{{{Hello there!}}}


**{variable : .xf}**

In [87]:
price = 1257425.32568
print(f"The price of the Product is ${price:}")
print(f"The price of the Product is ${price:.0f}")
print(f"The price of the Product is ${price:,}")
print(f"The price of the Product is ${price:,.2f}")
print(f"The price of the Product is ${price:_}")

The price of the Product is $1257425.32568
The price of the Product is $1257425
The price of the Product is $1,257,425.32568
The price of the Product is $1,257,425.33
The price of the Product is $1_257_425.32568


**In python, `.2f` is a format specifier used within string formatting to display a floating-point number**

In [88]:
info = {'Id' : 101, 'Name' : 'Aliya'}

print(f"Id of {info['Name']} is {info['Id']}")

Id of Aliya is 101


In [89]:
d = (9, 6, 2025)
print(f"Date : {d[0]}-{d[1]}-{d[2]}")
print(f"Date : {d[0]:02}-{d[1]:02}-{d[2]}")
print(f"Date : {d[0]:02}/{d[1]:02}/{d[2]}")

Date : 9-6-2025
Date : 09-06-2025
Date : 09/06/2025


In [90]:
words = ["Hello", "World!", "I", "am", "a", "Human!"]

In [91]:
f"{'\n'.join(words)}"

'Hello\nWorld!\nI\nam\na\nHuman!'

In [92]:
print(f"{'\n'.join(words)}")

Hello
World!
I
am
a
Human!


In [93]:
print(repr(f"{'\n'.join(words)}"))

'Hello\nWorld!\nI\nam\na\nHuman!'


#### Using str.format()

In [94]:
#using keyword arguments
print("Name: {name}, Age:{age}".format(name="Aliya", age=25))

Name: Aliya, Age:25


In [95]:
#classic method
name = "Aliya"
age = 25
print("Hello, {}! You are {} years old".format(name, age))

Hello, Aliya! You are 25 years old


In [96]:
#using positional arguments
print("Name: {0}, Age: {1}".format(name, age))

Name: Aliya, Age: 25


In [97]:
print("Name: {1}, Age: {0}".format(name, age))

Name: 25, Age: Aliya


In [98]:
print("The order is {0}, {1} and another {0}.".format("apple","banana"))

The order is apple, banana and another apple.


#### Using Format specifier(%)

**Format Specifiers** or the **modulo (%) operator** for string formatting is the oldest method in Python.\
Most commonly used format specifiers:\
**%s** - string\
**%c** - single character\
**%d** - integer (decimal)\
**%f** - float\
**%(sp).(num)f** - float... (sp) denotes no. of space before no. and (num) denotes no. of digits after .\

In [99]:
a = "This is a string"
print("String is '%s'" % a)

String is 'This is a string'


In [100]:
names = ["Aliya", "Arman", "Yunus"]
print('The name of the first candidate is %s' %names[0])
print('The name of the second candidate is %s' %names[1])
print('The name of the third candidate is %s' %names[2])

The name of the first candidate is Aliya
The name of the second candidate is Arman
The name of the third candidate is Yunus


In [101]:
# single character
x = 'a'
print("Single character is %c" % x)

Single character is a


In [102]:
y = '1'
print("Single character is %c" % y)

Single character is 1


In [103]:
word = 'cat'
print("Second letter of the word is '%c'" % word[1])

Second letter of the word is 'a'


In [104]:
# integer
c = 45
print("number is %d" % c)

number is 45


In [105]:
# float without specified precision
d = 34.521094
print("float is %f" % d)
print("float is %.2f" % d)
print("float is '%.2f'" % d)

float is 34.521094
float is 34.52
float is '34.52'


#### *Some more format specifiers:*\
**%x / %X** - hexadecimal representation of the value\
**%o** - octal representation of the value\

In [106]:
x = 11
print("Hexadecimal value of %d is %X" % (x, x))
print("Octal value of %d is %o" % (x, x))

Hexadecimal value of 11 is B
Octal value of 11 is 13


**%e / %E (Scientific Notation):**\
The %e specifier formats a floating-point number in scientific notation (exponential notation) with a lowercase 'e' indicating the exponent.

In [107]:
m = 1234.56789
print("For the number 1234.56789")
print("The value in scientific e notation is: %e" % m)
print("The value in scientific e notation is: %.1e" % m)
print("The value in scientific e notation is: %.2e" % m)
print("The value in scientific e notation is: %.3e" % m)
print("The value in scientific e notation is: %.4e" % m)

For the number 1234.56789
The value in scientific e notation is: 1.234568e+03
The value in scientific e notation is: 1.2e+03
The value in scientific e notation is: 1.23e+03
The value in scientific e notation is: 1.235e+03
The value in scientific e notation is: 1.2346e+03


In [108]:
n = 0.0000123456
print("\nFor the number 0.0000123456")
print("The value in scientific e notation is: %e" % n)
print("The value in scientific e notation is: %.1e" % n)
print("The value in scientific e notation is: %.2e" % n)
print("The value in scientific e notation is: %.3e" % n)
print("The value in scientific e notation is: %.4e" % n)


For the number 0.0000123456
The value in scientific e notation is: 1.234560e-05
The value in scientific e notation is: 1.2e-05
The value in scientific e notation is: 1.23e-05
The value in scientific e notation is: 1.235e-05
The value in scientific e notation is: 1.2346e-05


**%g / %G (General Format):**\
The ```%g``` specifier provides a "general" format for floating-point numbers. It chooses the more compact representation between fixed-point (%f) and scientific notation (%e), depending on the magnitude of the number, and suppresses unnecessary trailing zeros.

In [109]:
# Using the % operator
p = 123.45678
print("Large number: 123.45678")
print("General format: %g" % p)
print("General format: %.1g" % p)
print("General format: %.2g" % p)
print("General format: %.3g" % p)
print("General format: %.4g" % p)
print("General format: %.5g" % p)

Large number: 123.45678
General format: 123.457
General format: 1e+02
General format: 1.2e+02
General format: 123
General format: 123.5
General format: 123.46


In [110]:
q = 0.0000123
print("\nSmall number: 0.0000123")
print("General format: %g" % q)
print("General format: %.1g" % q)
print("General format: %.2g" % q)
print("General format: %.3g" % q)
print("General format: %.4g" % q)


Small number: 0.0000123
General format: 1.23e-05
General format: 1e-05
General format: 1.2e-05
General format: 1.23e-05
General format: 1.23e-05


In [111]:
r = 123456789.0
print("\nVery Large number: 123456789.0")
print("General format: %g" % r)
print("General format: %.1g" % r)
print("General format: %.2g" % r)
print("General format: %.3g" % r)
print("General format: %.4g" % r)
print("General format: %.5g" % r)


Very Large number: 123456789.0
General format: 1.23457e+08
General format: 1e+08
General format: 1.2e+08
General format: 1.23e+08
General format: 1.235e+08
General format: 1.2346e+08


In [112]:
f_num = 0.023
print("Number: 0.023")
print("\nA small number in scientific e notation: %e" %f_num)
print("A small number in scientific e notation: %.1e" %f_num)
print("A small number in scientific e notation: %.2e" %f_num)
print("\nA small number in general format: %g" %f_num)
print("A small number in general format: %.1g" %f_num)
print("A small number in general format: %.2g" %f_num)
print("A small number in general format: %.3g" %f_num)

Number: 0.023

A small number in scientific e notation: 2.300000e-02
A small number in scientific e notation: 2.3e-02
A small number in scientific e notation: 2.30e-02

A small number in general format: 0.023
A small number in general format: 0.02
A small number in general format: 0.023
A small number in general format: 0.023


In [113]:
d_num = 120000000
print("Number: 120000000")
print("\nA small number in scientific e notation: %e" %d_num)
print("A small number in scientific e notation: %.1e" %d_num)
print("A small number in scientific e notation: %.2e" %d_num)
print("\nA small number in general format: %g" %d_num)
print("A small number in general format: %.1g" %d_num)
print("A small number in general format: %.2g" %d_num)
print("A small number in general format: %.3g" %d_num)

Number: 120000000

A small number in scientific e notation: 1.200000e+08
A small number in scientific e notation: 1.2e+08
A small number in scientific e notation: 1.20e+08

A small number in general format: 1.2e+08
A small number in general format: 1e+08
A small number in general format: 1.2e+08
A small number in general format: 1.2e+08


***scientific e notation** is a way to express very large or very small numbers concisely. It replaces 'x10^' part of stantard notation with 'e'
For eg:\
the number 120000000 i.e 12crores becomes \
'1.2 x 10^8' as per scientific notation\
'1.2e8' as per scientific e notation\
\
similarly;\
the number 0.0000012 becomes\
'1.2 x 10^-6' as per scientific notification\
'1.2e-6' as per scientific e notification*